PropertyGraph: propagate loaded vs unloaded property info to views

tylershunt opened the pull request:
In my earlier change I missed these changes, and it turns out cases
where reasoning about what's loaded and what's not is important though
these classes.

Enterprise fix is in https://github.com/KatanaGraph/katana-enterprise/pull/1701

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

l-hoang commented on libquery/src/KFHandler.cpp:
```diff
@@ -12,7 +12,11 @@
 
-namespace galois {
+// defining some typedefs for use to make code less verbose
+using ArrowSchemaPointer       = std::shared_ptr<arrow::Schema>;
```
wrapping this in an anonymous namespace wouldn't do anything i assume

ddn0 replied:
Technically no but if you think that would be more natural I can move them down...

l-hoang replied:
nah, it's fine like this

i only learned about anonymous namespaces in .cpp files recently so i still don't know how the work completely

l-hoang commented on libquery/src/Table.cpp:
```diff
@@ -91,122 +96,115 @@ auto Untemplate(galois::Table::Type dtype, Args&&... args) {
 }
 
-galois::Table::Builder::Builder(const galois::Table::Schema& s,
+galois::Table::Builder::Builder(galois::Table::Schema schema,
```
would doing ``Schema&&`` make it more explicit that it's being moved

ddn0 replied:
There are two ways to indicate that you are making a copy (in a constructor):
```
T(const U& u): u_(u) {} // Pass by reference then copy
```
or 
```
T(U u): u_(std::move(u)) {} // Pass by copy then move
```

In both cases, only one copy is made. The second version has an additional move but also transparently supports move semantics like `T(std::move(old_u))` automatically where the first one would require another overload to handle this case.

l-hoang replied:
oh, this is for a copy, not a move; i see, thanks

l-hoang commented on libquery/src/Table.cpp:
```diff
@@ -271,15 +270,16 @@ std::shared_ptr<galois::Table> galois::Table::Builder::Make() {
       table_->schema()->field(column_number)->type()));
   std::shared_ptr<arrow::ChunkedArray> chunked = table_->column(column_number);
-  std::shared_ptr<arrow::Array> array          = chunked->chunk(0);
+  const std::shared_ptr<arrow::Array>& array   = chunked->chunks()[0];
```
i'm curious if this changed performance of the benchmark from yesterday since you're not incrementing a reference anymore

ddn0 replied:
maybe, though I find with locks, it isn't until you remove all of them do you really get good performance.

roshandathathri commented on libquery/src/MatchOp.cpp:
```diff
@@ -79,16 +79,18 @@ std::vector<galois::Operator*> galois::MatchOp::Compute() {
+    node_names.emplace_back(TableNodeName{
+        .name    = name,
+        .node_id = embedding_node_id,
+    });
```
Why not define a constructor for the struct? Wouldn't it make this code simpler?

ddn0 replied:
At least for me, there are basic POD types, where one deals with the data directly, which I find easiest to use in C-style (like `TableNodeName` here). And C++ classes, where one deals with methods and constructors, etc.

roshandathathri replied:
Ok. I was just curious.

roshandathathri commented on libquery/src/Table.cpp:
```diff
@@ -91,125 +96,119 @@ auto Untemplate(galois::Table::Type dtype, Args&&... args) {
-    return galois::Table::Builder(galois::Table::Schema());
+
+  if (embeddings.size() == 0) {
+    galois::Table::Builder builder{galois::Table::Schema()};
```
Is move semantics used here? (related to your comment above about the constructor)

ddn0 replied:
There isn't a reason to move a temporary variable:

https://godbolt.org/z/snjMT9

roshandathathri replied:
Ok, makes sense.

roshandathathri replied:
Why have the constructor allow move then? Were you targeting a specific use or it is just better in general?

ddn0 replied:
https://stackoverflow.com/questions/51705967/advantages-of-pass-by-value-and-stdmove-over-pass-by-reference/51706522

https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html

https://stackoverflow.com/questions/21035417/is-the-pass-by-value-and-then-move-construct-a-bad-idiom

roshandathathri replied:
Thanks for the clarification!

Commit:
PropertyGraph: propagate loaded vs unloaded property info to views

In my earlier change I missed these changes, and it turns out cases
where reasoning about what's loaded and what's not is important though
these classes.

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

witchel reviewed (approved):
Sets the stage.
